File: exercise/java/day15/src/test/java/santaChristmasList/BusinessTests.java
package santaChristmasList;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import santaChristmasList.operations.Business;
import santaChristmasList.operations.dependencies.Factory;
import santaChristmasList.operations.dependencies.Inventory;
import santaChristmasList.operations.dependencies.WishList;
import santaChristmasList.operations.models.Child;
import santaChristmasList.operations.models.Gift;
import santaChristmasList.operations.models.ManufacturedGift;

import static org.assertj.core.api.Assertions.assertThat;

public class BusinessTests {

    private static final String BAR_CODE = "123";

    private Factory factory;
    private Inventory inventory;
    private WishList wishList;

    private Child john;
    private Gift toy;
    private ManufacturedGift manufacturedGift;

    @BeforeEach
    public void setup() {
        factory = new Factory();
        inventory = new Inventory();
        wishList = new WishList();

        john = new Child("John");
        toy = new Gift("Toy");
        manufacturedGift = new ManufacturedGift(BAR_CODE);
    }

    @Test
    public void gift_ShouldBeLoadedIntoSleigh() {
        wishList.put(john, toy);
        factory.put(toy, manufacturedGift);
        inventory.put(BAR_CODE, toy);

        var business = new Business(factory, inventory, wishList);
        var sleigh = business.loadGiftsInSleigh(john);

        assertThat(sleigh.get(john)).isEqualTo("Gift: Toy has been loaded!");
    }

    @Test
    public void gift_ShouldNotBeLoaded_GivenChildIsNotOnWishList() {
        var business = new Business(factory, inventory, wishList);
        var sleigh = business.loadGiftsInSleigh(john);

        assertThat(sleigh.containsKey(john)).isFalse();
    }

    @Test
    public void gift_ShouldNotBeLoaded_GivenToyWasNotManufactured() {
        wishList.put(john, toy);

        var business = new Business(factory, inventory, wishList);
        var sleigh = business.loadGiftsInSleigh(john);

        assertThat(sleigh.containsKey(john)).isFalse();
    }

    @Test
    public void gift_ShouldNotBeLoaded_GivenToyWasMisplaced() {
        wishList.put(john, toy);
        factory.put(toy, manufacturedGift);

        var business = new Business(factory, inventory, wishList);
        var sleigh = business.loadGiftsInSleigh(john);

        assertThat(sleigh.containsKey(john)).isFalse();
    }
}

File: exercise/kotlin/day15/src/test/kotlin/santaChristmasList/BusinessTests.kt
package santaChristmasList

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.maps.shouldNotContainKey
import io.kotest.matchers.shouldBe
import santaChristmasList.operations.*

class BusinessTests : StringSpec({
    val barCode = "123"

    lateinit var factory: Factory
    lateinit var inventory: Inventory
    lateinit var wishList: WishList

    val john = Child("John")
    val toy = Gift("Toy")
    val manufacturedGift = ManufacturedGift(barCode)

    beforeTest {
        factory = Factory()
        inventory = Inventory()
        wishList = WishList()
    }

    "Gift should be loaded into Sleigh" {
        wishList[john] = toy
        factory[toy] = manufacturedGift
        inventory[barCode] = toy

        val business = Business(factory, inventory, wishList)
        val sleigh = business.loadGiftsInSleigh(john)

        sleigh[john] shouldBe "Gift: Toy has been loaded!"
    }

    "Gift should not be loaded given child is not on wish list" {
        val business = Business(factory, inventory, wishList)
        val sleigh = business.loadGiftsInSleigh(john)

        sleigh shouldNotContainKey john
    }

    "Gift should not be loaded given toy was not manufactured" {
        wishList[john] = toy

        val business = Business(factory, inventory, wishList)
        val sleigh = business.loadGiftsInSleigh(john)

        sleigh shouldNotContainKey john
    }

    "Gift should not be loaded given toy was misplaced" {
        wishList[john] = toy
        factory[toy] = manufacturedGift

        val business = Business(factory, inventory, wishList)
        val sleigh = business.loadGiftsInSleigh(john)

        sleigh shouldNotContainKey john
    }
})
